File: src/main/java/com/owncloud/android/ui/dialog/ChooseTemplateDialogFragment.java
/*
 * Nextcloud Android client application
 *
 * @author Tobias Kaminsky
 * @author Chris Narkiewicz
 *
 * Copyright (C) 2018 Tobias Kaminsky
 * Copyright (C) 2018 Nextcloud GmbH.
 * Copyright (C) 2019 Chris Narkiewicz <hello@ezaquarii.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package com.owncloud.android.ui.dialog;

import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.view.WindowManager.LayoutParams;
import android.widget.Button;

import com.nextcloud.android.lib.resources.directediting.DirectEditingCreateFileRemoteOperation;
import com.nextcloud.android.lib.resources.directediting.DirectEditingObtainListOfTemplatesRemoteOperation;
import com.nextcloud.client.account.CurrentAccountProvider;
import com.nextcloud.client.account.User;
import com.nextcloud.client.di.Injectable;
import com.nextcloud.client.network.ClientFactory;
import com.owncloud.android.MainApp;
import com.owncloud.android.R;
import com.owncloud.android.databinding.ChooseTemplateBinding;
import com.owncloud.android.datamodel.FileDataStorageManager;
import com.owncloud.android.datamodel.OCFile;
import com.owncloud.android.lib.common.Creator;
import com.owncloud.android.lib.common.OwnCloudClient;
import com.owncloud.android.lib.common.Template;
import com.owncloud.android.lib.common.TemplateList;
import com.owncloud.android.lib.common.operations.RemoteOperationResult;
import com.owncloud.android.lib.common.utils.Log_OC;
import com.owncloud.android.lib.resources.files.ReadFileRemoteOperation;
import com.owncloud.android.lib.resources.files.model.RemoteFile;
import com.owncloud.android.ui.activity.ExternalSiteWebView;
import com.owncloud.android.ui.activity.TextEditorWebView;
import com.owncloud.android.ui.adapter.TemplateAdapter;
import com.owncloud.android.utils.DisplayUtils;
import com.owncloud.android.utils.FileStorageUtils;
import com.owncloud.android.utils.theme.ThemeButtonUtils;
import com.owncloud.android.utils.theme.ThemeColorUtils;
import com.owncloud.android.utils.theme.ThemeTextInputUtils;

import java.lang.ref.WeakReference;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;
import androidx.recyclerview.widget.GridLayoutManager;

/**
 * Dialog to show templates for new documents/spreadsheets/presentations.
 */
public class ChooseTemplateDialogFragment extends DialogFragment implements View.OnClickListener,
    TemplateAdapter.ClickListener, Injectable {

    private static final String ARG_PARENT_FOLDER = "PARENT_FOLDER";
    private static final String ARG_CREATOR = "CREATOR";
    private static final String ARG_HEADLINE = "HEADLINE";
    private static final String TAG = ChooseTemplateDialogFragment.class.getSimpleName();
    private static final String DOT = ".";
    public static final int SINGLE_TEMPLATE = 1;

    private TemplateAdapter adapter;
    private OCFile parentFolder;
    private String title;
    @Inject ClientFactory clientFactory;
    private Creator creator;
    @Inject CurrentAccountProvider currentAccount;
    private Button positiveButton;

    public enum Type {
        DOCUMENT,
        SPREADSHEET,
        PRESENTATION
    }

    ChooseTemplateBinding binding;

    public static ChooseTemplateDialogFragment newInstance(OCFile parentFolder, Creator creator, String headline) {
        ChooseTemplateDialogFragment frag = new ChooseTemplateDialogFragment();
        Bundle args = new Bundle();
        args.putParcelable(ARG_PARENT_FOLDER, parentFolder);
        args.putParcelable(ARG_CREATOR, creator);
        args.putString(ARG_HEADLINE, headline);
        frag.setArguments(args);
        return frag;
    }

    @Override
    public void onStart() {
        super.onStart();

        AlertDialog alertDialog = (AlertDialog) getDialog();

        positiveButton = alertDialog.getButton(AlertDialog.BUTTON_POSITIVE);
        ThemeButtonUtils.themeBorderlessButton(positiveButton, alertDialog.getButton(AlertDialog.BUTTON_NEUTRAL));
        positiveButton.setOnClickListener(this);
        positiveButton.setEnabled(false);

        checkEnablingCreateButton();
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Bundle arguments = getArguments();
        if (arguments == null) {
            throw new IllegalArgumentException("Arguments may not be null");
        }

        Activity activity = getActivity();
        if (activity == null) {
            throw new IllegalArgumentException("Activity may not be null");
        }

        parentFolder = arguments.getParcelable(ARG_PARENT_FOLDER);
        creator = arguments.getParcelable(ARG_CREATOR);
        title = arguments.getString(ARG_HEADLINE, getString(R.string.select_template));

        if (savedInstanceState == null) {
            title = arguments.getString(ARG_HEADLINE);
        } else {
            title = savedInstanceState.getString(ARG_HEADLINE);
        }

        // Inflate the layout for the dialog
        LayoutInflater inflater = requireActivity().getLayoutInflater();
        binding = ChooseTemplateBinding.inflate(inflater, null, false);
        View view = binding.getRoot();

        binding.filename.requestFocus();
        ThemeTextInputUtils.colorTextInput(binding.filenameContainer,
                                           binding.filename,
                                           ThemeColorUtils.primaryColor(getContext()));

        binding.filename.setOnKeyListener((v, keyCode, event) -> {
            checkEnablingCreateButton();
            return false;
        });

        binding.filename.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                // generated method stub
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // generated method stub
            }

            @Override
            public void afterTextChanged(Editable s) {
                checkEnablingCreateButton();
            }
        });

        try {
            User user = currentAccount.getUser();
            new FetchTemplateTask(this, clientFactory, user, creator).execute();
        } catch (Exception e) {
            Log_OC.e(TAG, "Loading stream url not possible: " + e);
        }

        binding.list.setHasFixedSize(true);
        binding.list.setLayoutManager(new GridLayoutManager(activity, 2));
        adapter = new TemplateAdapter(creator.getMimetype(), this, getContext(), currentAccount, clientFactory);
        binding.list.setAdapter(adapter);

        // Build the dialog
        AlertDialog.Builder builder = new AlertDialog.Builder(activity);
        builder.setView(view)
            .setPositiveButton(R.string.create, null)
            .setNeutralButton(R.string.common_cancel, null)
            .setTitle(title);
        Dialog dialog = builder.create();

        Window window = dialog.getWindow();

        if (window != null) {
            window.setSoftInputMode(LayoutParams.SOFT_INPUT_STATE_VISIBLE);
        }

        return dialog;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle savedInstanceState) {
        super.onSaveInstanceState(savedInstanceState);
        savedInstanceState.putString(ARG_HEADLINE, title);
    }

    private void createFromTemplate(Template template, String path) {
        new CreateFileFromTemplateTask(this, clientFactory, currentAccount.getUser(), template, path, creator).execute();
    }

    public void setTemplateList(TemplateList templateList) {
        adapter.setTemplateList(templateList);
        adapter.notifyDataSetChanged();
    }

    @Override
    public void onClick(Template template) {
        onTemplateChosen(template);
    }

    private void onTemplateChosen(Template template) {
        adapter.setTemplateAsActive(template);
        prefillFilenameIfEmpty(template);
        checkEnablingCreateButton();
    }

    private void prefillFilenameIfEmpty(Template template) {
        String name = binding.filename.getText().toString();
        if (name.isEmpty() || name.equalsIgnoreCase(DOT + template.getExtension())) {
            binding.filename.setText(String.format("%s.%s", template.title, template.extension));
        }
        binding.filename.setSelection(binding.filename.getText().toString().lastIndexOf('.'));
    }

    @Override
    public void onClick(View v) {
        String name = binding.filename.getText().toString();
        String path = parentFolder.getRemotePath() + name;

        Template selectedTemplate = adapter.getSelectedTemplate();

        if (selectedTemplate == null) {
            DisplayUtils.showSnackMessage(binding.list, R.string.select_one_template);
        } else if (name.isEmpty() || name.equalsIgnoreCase(DOT + selectedTemplate.getExtension())) {
            DisplayUtils.showSnackMessage(binding.list, R.string.enter_filename);
        } else if (!name.endsWith(selectedTemplate.getExtension())) {
            createFromTemplate(selectedTemplate, path + DOT + selectedTemplate.getExtension());
        } else {
            createFromTemplate(selectedTemplate, path);
        }
    }

    private void checkEnablingCreateButton() {
        Template selectedTemplate = adapter.getSelectedTemplate();
        String name = binding.filename.getText().toString();

        positiveButton.setEnabled(selectedTemplate != null && !name.isEmpty() &&
                                      !name.equalsIgnoreCase(DOT + selectedTemplate.getExtension()));
    }

    private static class CreateFileFromTemplateTask extends AsyncTask<Void, Void, String> {
        private ClientFactory clientFactory;
        private WeakReference<ChooseTemplateDialogFragment> chooseTemplateDialogFragmentWeakReference;
        private Template template;
        private String path;
        private Creator creator;
        private User user;
        private OCFile file;

        CreateFileFromTemplateTask(ChooseTemplateDialogFragment chooseTemplateDialogFragment,
                                   ClientFactory clientFactory,
                                   User user,
                                   Template template,
                                   String path,
                                   Creator creator
        ) {
            this.clientFactory = clientFactory;
            this.chooseTemplateDialogFragmentWeakReference = new WeakReference<>(chooseTemplateDialogFragment);
            this.template = template;
            this.path = path;
            this.creator = creator;
            this.user = user;
        }

        @Override
        protected String doInBackground(Void... voids) {
            try {
                OwnCloudClient client = clientFactory.create(user);

                RemoteOperationResult<String> result =
                    new DirectEditingCreateFileRemoteOperation(path,
                                                               creator.getEditor(),
                                                               creator.getId(),
                                                               template.getTitle()).execute(client);
                if (!result.isSuccess()) {
                    return "";
                }

                RemoteOperationResult newFileResult = new ReadFileRemoteOperation(path).execute(client);
                if (!newFileResult.isSuccess()) {
                    return "";
                }

                final ChooseTemplateDialogFragment fragment = chooseTemplateDialogFragmentWeakReference.get();
                if (fragment == null) {
                    return "";
                }

                final Context context = fragment.getContext();
                if (context == null) {
                    // fragment has been detached
                    return "";
                }

                FileDataStorageManager storageManager = new FileDataStorageManager(user.toPlatformAccount(),
                                                                                   context.getContentResolver());

                OCFile temp = FileStorageUtils.fillOCFile((RemoteFile) newFileResult.getData().get(0));
                storageManager.saveFile(temp);
                file = storageManager.getFileByPath(path);

                return result.getResultData();

            } catch (ClientFactory.CreationException e) {
                Log_OC.e(TAG, "Error creating file from template!", e);
                return "";
            }
        }

        @Override
        protected void onPostExecute(String url) {
            final ChooseTemplateDialogFragment fragment = chooseTemplateDialogFragmentWeakReference.get();

            if (fragment != null && fragment.isAdded()) {
                if (url.isEmpty()) {
                    DisplayUtils.showSnackMessage(fragment.binding.list, "Error creating file from template");
                } else {
                    Intent editorWebView = new Intent(MainApp.getAppContext(), TextEditorWebView.class);
                    editorWebView.putExtra(ExternalSiteWebView.EXTRA_TITLE, "Text");
                    editorWebView.putExtra(ExternalSiteWebView.EXTRA_URL, url);
                    editorWebView.putExtra(ExternalSiteWebView.EXTRA_FILE, file);
                    editorWebView.putExtra(ExternalSiteWebView.EXTRA_SHOW_SIDEBAR, false);
                    fragment.startActivity(editorWebView);

                    fragment.dismiss();
                }
            } else {
                Log_OC.e(TAG, "Error creating file from template!");
            }
        }
    }

    private static class FetchTemplateTask extends AsyncTask<Void, Void, TemplateList> {

        private User user;
        private ClientFactory clientFactory;
        private WeakReference<ChooseTemplateDialogFragment> chooseTemplateDialogFragmentWeakReference;
        private Creator creator;

        FetchTemplateTask(ChooseTemplateDialogFragment chooseTemplateDialogFragment,
                          ClientFactory clientFactory,
                          User user,
                          Creator creator) {
            this.user = user;
            this.clientFactory = clientFactory;
            this.chooseTemplateDialogFragmentWeakReference = new WeakReference<>(chooseTemplateDialogFragment);
            this.creator = creator;
        }

        @Override
        protected TemplateList doInBackground(Void... voids) {

            try {
                OwnCloudClient client = clientFactory.create(user);
                RemoteOperationResult<TemplateList> result =
                    new DirectEditingObtainListOfTemplatesRemoteOperation(creator.getEditor(),
                                                                          creator.getId())
                        .execute(client);

                if (!result.isSuccess()) {
                    return new TemplateList();
                }

                return result.getResultData();
            } catch (ClientFactory.CreationException e) {
                Log_OC.e(TAG, "Could not fetch template", e);

                return new TemplateList();
            }
        }

        @Override
        protected void onPostExecute(TemplateList templateList) {
            ChooseTemplateDialogFragment fragment = chooseTemplateDialogFragmentWeakReference.get();

            if (fragment != null && fragment.isAdded()) {
                if (templateList.templates.isEmpty()) {
                    DisplayUtils.showSnackMessage(fragment.binding.list, R.string.error_retrieving_templates);
                } else {
                    if (templateList.templates.size() == SINGLE_TEMPLATE) {
                        fragment.onTemplateChosen(templateList.templates.values().iterator().next());
                        fragment.binding.list.setVisibility(View.GONE);
                    } else {
                        String name = DOT + templateList.templates.values().iterator().next().getExtension();
                        fragment.binding.filename.setText(name);
                        fragment.binding.helperText.setVisibility(View.VISIBLE);
                    }

                    fragment.setTemplateList(templateList);
                }
            } else {
                Log_OC.e(TAG, "Error streaming file: no previewMediaFragment!");
            }
        }
    }
}

File: app/src/main/java/com/owncloud/android/ui/dialog/ChooseTemplateDialogFragment.kt
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2024 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-FileCopyrightText: 2023 TSI-mc
 * SPDX-FileCopyrightText: 2022 Álvaro Brey <alvaro.brey@nextcloud.com>
 * SPDX-FileCopyrightText: 2019 Chris Narkiewicz <hello@ezaquarii.com>
 * SPDX-FileCopyrightText: 2018 Tobias Kaminsky <tobias@kaminsky.me>
 * SPDX-FileCopyrightText: 2018 Nextcloud GmbH
 * SPDX-License-Identifier: GPL-3.0-or-later AND AGPL-3.0-or-later
 */
package com.owncloud.android.ui.dialog

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.Intent
import android.os.AsyncTask
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import androidx.recyclerview.widget.GridLayoutManager
import com.google.android.material.button.MaterialButton
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.nextcloud.android.lib.resources.directediting.DirectEditingCreateFileRemoteOperation
import com.nextcloud.android.lib.resources.directediting.DirectEditingObtainListOfTemplatesRemoteOperation
import com.nextcloud.client.account.CurrentAccountProvider
import com.nextcloud.client.account.User
import com.nextcloud.client.di.Injectable
import com.nextcloud.client.network.ClientFactory
import com.nextcloud.client.network.ClientFactory.CreationException
import com.nextcloud.utils.extensions.getParcelableArgument
import com.nextcloud.utils.fileNameValidator.FileNameValidator
import com.owncloud.android.MainApp
import com.owncloud.android.R
import com.owncloud.android.databinding.ChooseTemplateBinding
import com.owncloud.android.datamodel.FileDataStorageManager
import com.owncloud.android.datamodel.OCFile
import com.owncloud.android.lib.common.Creator
import com.owncloud.android.lib.common.Template
import com.owncloud.android.lib.common.TemplateList
import com.owncloud.android.lib.common.utils.Log_OC
import com.owncloud.android.lib.resources.files.ReadFileRemoteOperation
import com.owncloud.android.lib.resources.files.model.RemoteFile
import com.owncloud.android.lib.resources.status.OCCapability
import com.owncloud.android.ui.activity.ExternalSiteWebView
import com.owncloud.android.ui.activity.TextEditorWebView
import com.owncloud.android.ui.adapter.TemplateAdapter
import com.owncloud.android.utils.DisplayUtils
import com.owncloud.android.utils.FileStorageUtils
import com.owncloud.android.utils.KeyboardUtils
import com.owncloud.android.utils.theme.ViewThemeUtils
import java.lang.ref.WeakReference
import javax.inject.Inject

/**
 * Dialog to show templates for new documents/spreadsheets/presentations.
 */
class ChooseTemplateDialogFragment :
    DialogFragment(),
    View.OnClickListener,
    TemplateAdapter.ClickListener,
    Injectable {

    private lateinit var fileNames: MutableSet<String>

    @Inject
    lateinit var clientFactory: ClientFactory

    @Inject
    lateinit var currentAccount: CurrentAccountProvider

    @Inject
    lateinit var fileDataStorageManager: FileDataStorageManager

    @Inject
    lateinit var viewThemeUtils: ViewThemeUtils

    @Inject
    lateinit var keyboardUtils: KeyboardUtils

    private var adapter: TemplateAdapter? = null
    private var parentFolder: OCFile? = null
    private var title: String? = null
    private var positiveButton: MaterialButton? = null
    private var creator: Creator? = null

    private var _binding: ChooseTemplateBinding? = null
    val binding get() = _binding!!

    override fun onStart() {
        super.onStart()
        val alertDialog = dialog as AlertDialog

        val negativeButton = alertDialog.getButton(AlertDialog.BUTTON_NEGATIVE) as? MaterialButton
        negativeButton?.let {
            viewThemeUtils.material.colorMaterialButtonPrimaryBorderless(negativeButton)
        }

        val positiveButton = alertDialog.getButton(AlertDialog.BUTTON_POSITIVE) as? MaterialButton
        positiveButton?.let {
            viewThemeUtils.material.colorMaterialButtonPrimaryTonal(positiveButton)
            positiveButton.setOnClickListener(this)
            positiveButton.isEnabled = false
            positiveButton.isClickable = false
            this.positiveButton = positiveButton
        }

        checkFileNameAfterEachType()
    }

    override fun onResume() {
        super.onResume()
        keyboardUtils.showKeyboardForEditText(dialog?.window, binding.filename)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val arguments = arguments ?: throw IllegalArgumentException("Arguments may not be null")
        val activity = activity ?: throw IllegalArgumentException("Activity may not be null")

        parentFolder = arguments.getParcelableArgument(ARG_PARENT_FOLDER, OCFile::class.java)
        creator = arguments.getParcelableArgument(ARG_CREATOR, Creator::class.java)

        title = arguments.getString(ARG_HEADLINE, getString(R.string.select_template))
        title = when (savedInstanceState) {
            null -> arguments.getString(ARG_HEADLINE)
            else -> savedInstanceState.getString(ARG_HEADLINE)
        }

        fileNames = fileDataStorageManager.getFolderContent(parentFolder, false).map { it.fileName }.toMutableSet()

        val inflater = requireActivity().layoutInflater
        _binding = ChooseTemplateBinding.inflate(inflater, null, false)

        viewThemeUtils.material.colorTextInputLayout(
            binding.filenameContainer
        )

        binding.filename.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) = Unit
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) = Unit
            override fun afterTextChanged(s: Editable) {
                checkFileNameAfterEachType()
            }
        })

        fetchTemplate()

        binding.list.setHasFixedSize(true)
        binding.list.layoutManager = GridLayoutManager(activity, 2)
        adapter = TemplateAdapter(
            creator?.mimetype,
            this,
            context,
            currentAccount,
            clientFactory,
            viewThemeUtils
        )
        binding.list.adapter = adapter

        val builder = MaterialAlertDialogBuilder(activity)
            .setView(binding.root)
            .setPositiveButton(R.string.create, null)
            .setNegativeButton(R.string.common_cancel, null)
            .setTitle(title)

        viewThemeUtils.dialog.colorMaterialAlertDialogBackground(binding.list.context, builder)

        return builder.create()
    }

    @Suppress("TooGenericExceptionCaught")
    private fun fetchTemplate() {
        try {
            val user = currentAccount.user
            FetchTemplateTask(this, clientFactory, user, creator).execute()
        } catch (e: Exception) {
            Log_OC.e(TAG, "Loading stream url not possible: $e")
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onSaveInstanceState(savedInstanceState: Bundle) {
        super.onSaveInstanceState(savedInstanceState)
        savedInstanceState.putString(ARG_HEADLINE, title)
    }

    private fun createFromTemplate(template: Template, path: String) {
        CreateFileFromTemplateTask(this, clientFactory, currentAccount.user, template, path, creator).execute()
    }

    @SuppressLint("NotifyDataSetChanged")
    fun setTemplateList(templateList: TemplateList?) {
        adapter?.setTemplateList(templateList)
        adapter?.notifyDataSetChanged()
    }

    override fun onClick(template: Template) {
        onTemplateChosen(template)
    }

    private fun onTemplateChosen(template: Template) {
        adapter?.setTemplateAsActive(template)
        prefillFilenameIfEmpty(template)
        checkFileNameAfterEachType()
    }

    private fun prefillFilenameIfEmpty(template: Template) {
        var name = binding.filename.text.toString()
        if (name.isEmpty() || name.equals(DOT + template.extension, ignoreCase = true)) {
            binding.filename.setText(String.format("%s.%s", template.title, template.extension))
            name = binding.filename.text.toString()
            val dotPos = name.lastIndexOf('.')
            binding.filename.setSelection(if (dotPos != -1) dotPos else name.length)
        }
    }

    private fun getOCCapability(): OCCapability = fileDataStorageManager.getCapability(currentAccount.user.accountName)

    override fun onClick(v: View) {
        val name = binding.filename.text.toString()
        val path = parentFolder?.remotePath + name
        val selectedTemplate = adapter?.selectedTemplate

        val errorMessage = FileNameValidator.checkFileName(name, getOCCapability(), requireContext())

        when {
            selectedTemplate == null -> {
                DisplayUtils.showSnackMessage(binding.list, R.string.select_one_template)
            }
            errorMessage != null -> {
                DisplayUtils.showSnackMessage(requireActivity(), errorMessage)
            }
            name.equals(DOT + selectedTemplate.extension, ignoreCase = true) -> {
                DisplayUtils.showSnackMessage(binding.list, R.string.enter_filename)
            }
            else -> {
                val fullPath = if (!name.endsWith(selectedTemplate.extension)) {
                    path + DOT + selectedTemplate.extension
                } else {
                    path
                }
                createFromTemplate(selectedTemplate, fullPath)
            }
        }
    }

    private fun checkFileNameAfterEachType() {
        if (positiveButton == null) return

        val selectedTemplate = adapter?.selectedTemplate
        val name = binding.filename.text.toString().trim()
        val isNameJustExtension = selectedTemplate != null &&
            name.equals(
                DOT + selectedTemplate.extension,
                ignoreCase = true
            )
        val fileNameValidatorResult =
            FileNameValidator.checkFileName(name, getOCCapability(), requireContext(), fileNames)

        val errorMessage = when {
            isNameJustExtension -> null
            fileNameValidatorResult != null -> fileNameValidatorResult
            else -> null
        }

        val isNameValid = (errorMessage == null) && !name.equals(DOT + selectedTemplate?.extension, ignoreCase = true)
        val isHiddenFileName = FileNameValidator.isFileHidden(name)

        binding.filenameContainer.isErrorEnabled = !isNameValid || isHiddenFileName
        binding.filenameContainer.error = when {
            !isNameValid -> errorMessage ?: getString(R.string.enter_filename)
            isHiddenFileName -> getText(R.string.hidden_file_name_warning)
            else -> null
        }

        positiveButton?.apply {
            isEnabled = isNameValid && !isHiddenFileName
            isClickable = isEnabled
        }
    }

    @Suppress("LongParameterList", "DEPRECATION")
    private class CreateFileFromTemplateTask(
        chooseTemplateDialogFragment: ChooseTemplateDialogFragment,
        private val clientFactory: ClientFactory?,
        private val user: User,
        private val template: Template,
        private val path: String,
        private val creator: Creator?
    ) : AsyncTask<Void, Void, String>() {
        private val chooseTemplateDialogFragmentWeakReference: WeakReference<ChooseTemplateDialogFragment> =
            WeakReference(chooseTemplateDialogFragment)
        private var file: OCFile? = null

        @Deprecated("Deprecated in Java")
        @Suppress("ReturnCount") // legacy code
        override fun doInBackground(vararg params: Void): String {
            return try {
                val client = clientFactory?.create(user) ?: return ""
                val nextcloudClient = clientFactory.createNextcloudClient(user)
                val result = DirectEditingCreateFileRemoteOperation(
                    path,
                    creator?.editor,
                    creator?.id,
                    template.title
                ).execute(nextcloudClient)
                if (!result.isSuccess) {
                    return ""
                }
                val newFileResult = ReadFileRemoteOperation(path).execute(client)
                if (!newFileResult.isSuccess) {
                    return ""
                }
                val fragment = chooseTemplateDialogFragmentWeakReference.get() ?: return ""
                val context = fragment.context
                    ?: // fragment has been detached
                    return ""
                val storageManager = FileDataStorageManager(
                    user,
                    context.contentResolver
                )
                val temp = FileStorageUtils.fillOCFile(newFileResult.data[0] as RemoteFile)
                storageManager.saveFile(temp)
                file = storageManager.getFileByPath(path)
                result.resultData
            } catch (e: CreationException) {
                Log_OC.e(TAG, "Error creating file from template!", e)
                ""
            }
        }

        @Deprecated("Deprecated in Java")
        override fun onPostExecute(url: String) {
            val fragment = chooseTemplateDialogFragmentWeakReference.get()
            if (fragment == null || !fragment.isAdded) {
                Log_OC.e(TAG, "Error creating file from template!")
                return
            }

            if (url.isEmpty()) {
                DisplayUtils.showSnackMessage(fragment.binding.list, R.string.error_creating_file_from_template)
                return
            }

            val editorWebView = Intent(MainApp.getAppContext(), TextEditorWebView::class.java).apply {
                putExtra(ExternalSiteWebView.EXTRA_TITLE, "Text")
                putExtra(ExternalSiteWebView.EXTRA_URL, url)
                putExtra(ExternalSiteWebView.EXTRA_FILE, file)
                putExtra(ExternalSiteWebView.EXTRA_SHOW_SIDEBAR, false)
            }

            fragment.run {
                startActivity(editorWebView)
                dismiss()
            }
        }
    }

    @Suppress("DEPRECATION")
    private class FetchTemplateTask(
        chooseTemplateDialogFragment: ChooseTemplateDialogFragment,
        private val clientFactory: ClientFactory?,
        private val user: User,
        private val creator: Creator?
    ) : AsyncTask<Void, Void, TemplateList>() {
        private val chooseTemplateDialogFragmentWeakReference: WeakReference<ChooseTemplateDialogFragment> =
            WeakReference(chooseTemplateDialogFragment)

        @Deprecated("Deprecated in Java")
        override fun doInBackground(vararg voids: Void): TemplateList {
            return try {
                val client = clientFactory?.createNextcloudClient(user) ?: return TemplateList()
                val result = DirectEditingObtainListOfTemplatesRemoteOperation(
                    creator?.editor,
                    creator?.id
                ).execute(client)
                if (!result.isSuccess) {
                    TemplateList()
                } else {
                    result.resultData
                }
            } catch (e: CreationException) {
                Log_OC.e(TAG, "Could not fetch template", e)
                TemplateList()
            }
        }

        @Deprecated("Deprecated in Java")
        override fun onPostExecute(templateList: TemplateList) {
            val fragment = chooseTemplateDialogFragmentWeakReference.get()
            if (fragment == null || !fragment.isAdded) {
                Log_OC.e(TAG, "Error streaming file: no previewMediaFragment!")
                return
            }

            if (templateList.templates.isEmpty()) {
                DisplayUtils.showSnackMessage(fragment.binding.list, R.string.error_retrieving_templates)
                return
            }

            fragment.run {
                if (templateList.templates.size == SINGLE_TEMPLATE) {
                    onTemplateChosen(templateList.templates.values.iterator().next())
                    binding.list.visibility = View.GONE
                } else {
                    val name = DOT + templateList.templates.values.iterator().next().extension
                    binding.filename.setText(name)
                    binding.helperText.visibility = View.VISIBLE
                }

                setTemplateList(templateList)
            }
        }
    }

    companion object {
        private const val ARG_PARENT_FOLDER = "PARENT_FOLDER"
        private const val ARG_CREATOR = "CREATOR"
        private const val ARG_HEADLINE = "HEADLINE"
        private val TAG = ChooseTemplateDialogFragment::class.java.simpleName
        private const val DOT = "."
        const val SINGLE_TEMPLATE = 1

        @JvmStatic
        fun newInstance(parentFolder: OCFile?, creator: Creator?, headline: String?): ChooseTemplateDialogFragment {
            val bundle = Bundle().apply {
                putParcelable(ARG_PARENT_FOLDER, parentFolder)
                putParcelable(ARG_CREATOR, creator)
                putString(ARG_HEADLINE, headline)
            }

            return ChooseTemplateDialogFragment().apply {
                arguments = bundle
            }
        }
    }
}
